File: WatchBefore/app/src/main/java/com/example/watchbefore/MainActivity.java
package com.example.watchbefore;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.TextView;

import java.util.Locale;

public class MainActivity extends AppCompatActivity {
    //스톱워치에 표시할 초
    private int seconds = 0;
    //스톱워치가 실행 중인가?
    //버튼에 대한 이벤트로 상태값이 바뀌고 사용은 다른 메소드에서 치리가 되니깐 이런 경우는 전역변수로 해야한다.
    //화면을 회전시키면 왜 액티비티를 다시 생성해야하지?
    //onCreate는 보통 화면을 설정하는 코드를 포함합니다.
    //onCreate메소드에서 수평, 수직모드와 관련된 특정 구성에 의존하고 있다.
    //bundle은 액티비티 상태를 저장하고 있으므로 안드로이드 os를 통해서 상태를 전달,관리할 수 있는 유용한 API이다.
    private boolean running;//false -> start버튼을 누르면 true로
    /* 지금까지의 흐른 시간과 현재 스톱워치가 실행 중인지 상태 저장*/
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //별동의 메소드를 이용해서 스톱워치를 갱신한다.
        //액티비티가 생성이되면 메소드를 시작합니다.
        runTimer();
    }
    public void start(View v){
        running = true;
    }
    public void stop(View v){
        running = false;
    }
    public void reset(View v){
        running = false;
        seconds = 0;
    }

    /*
    여기서는 레이아웃의 텍스트뷰 레퍼런스를 얻고 seconds변수의 값을 시, 분, 초
    형식으로 바꾼다음 텍스트뷰에 결과 표시
    만일 runing값이 true이면 seconds변수의 값을 1씩 증가시킬 것.
     */
    private void runTimer(){
        final TextView tv_time = findViewById(R.id.tv_time);
        //Handler사용하면 특정코드를 미래의 특정시점에 사용할 수 있음.
        final Handler handler = new Handler();
        handler.post(new Runnable(){
            @Override
            public void run() {
                int hours = seconds/3600;
                int minutes = (seconds%3600)/60;
                int secs = seconds%60;
                String time = String.format(Locale.getDefault()
                        , "%d:%02d:%02d",hours,minutes,secs);
                tv_time.setText(time);
                if(running){
                    seconds++;
                }
                handler.postDelayed(this,1000);
            }
        });
    }//runTimer
}

File: KTBook802/app/src/main/java/com/example/ktbook802/MainActivity.kt
package com.example.ktbook802

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.MotionEvent
import android.widget.LinearLayout
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.room.Room
import com.example.ktbook802.adapter.BookAdapter
import com.example.ktbook802.adapter.HistoryAdapter
import com.example.ktbook802.api.BookService
import com.example.ktbook802.databinding.ActivityMainBinding
import com.example.ktbook802.model.BestSellerDTO
import com.example.ktbook802.model.History
import com.example.ktbook802.model.SearchBookDTO
import com.example.ktbook802.util.AppDatabase
import retrofit2.*
import retrofit2.converter.gson.GsonConverterFactory


class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var adapter: BookAdapter
    private lateinit var historyAdapter: HistoryAdapter
    private lateinit var bookService: BookService
    //데이터 베이스 가져오기
    private lateinit var db: AppDatabase
    // : 는 extends를 의미. 클래스에 ()을 붙이는 것도 자바랑 다름!
    // fun은 function
    // ?가 있다면 null도 가능하다.
    // ?이 없다면 null-fafe(null이 불가능)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        adapter = BookAdapter()
        binding.bookRecyclerView.layoutManager = LinearLayoutManager(this)
        binding.bookRecyclerView.adapter = adapter
        initHistoryRecyclerView()
        db = Room.databaseBuilder(applicationContext
            ,AppDatabase::class.java
            ,"BookSearchDB").build()

        //래트로핏 구현체 생성
        val retrofit = Retrofit.Builder()
            .baseUrl("https://book.interpark.com")
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        bookService = retrofit.create(BookService::class.java)
        ////////////////////[[ 베스트 셀러 목록 시작 ]]///////////////////
        bookService.getBestSellerBooks(getString(R.string.bookApiKey))
            .enqueue(object: Callback<BestSellerDTO> {
                override fun onResponse(
                    call: Call<BestSellerDTO>,
                    response: Response<BestSellerDTO>
                ) {
                    if(response.isSuccessful.not()){
                        Log.e(TAG, "NOT!!! SUCCESS")
                        return;
                    }
                    adapter.submitList(response.body()?.books.orEmpty())
                }

                override fun onFailure(call: Call<BestSellerDTO>, t: Throwable) {
                    //실패처리
                    Log.e(TAG, t.toString())
                }
            })
        ////////////////////[[ 베스트 셀러 목록  끝  ]]///////////////////
        initSearchEditText()
    }//////////////end of onCreate
    private fun initSearchEditText(){
        binding.searchEditText.setOnKeyListener{v, keyCode, event ->
            if(keyCode == KeyEvent.KEYCODE_ENTER && event.action == MotionEvent.ACTION_DOWN){
                search(binding.searchEditText.text.toString())
                return@setOnKeyListener true
            }
            return@setOnKeyListener false
        }
        binding.searchEditText.setOnTouchListener{v, event ->
            if(event.action == MotionEvent.ACTION_DOWN){
                showHistoryView()
            }
            return@setOnTouchListener false
        }
    }
    //검색기능 구현
    private fun search(keyword: String){
        bookService.getBooksByName(getString(R.string.bookApiKey),keyword)
            .enqueue(object: Callback<SearchBookDTO> {
                override fun onResponse(
                    call: Call<SearchBookDTO>,
                    response: Response<SearchBookDTO>
                ) {
                    hideHistoryView()
                    saveSearchKeyword(keyword)//;insertHistory
                    if(response.isSuccessful.not()){
                        Log.e(TAG, "NOT!!! SUCCESS")
                        return;
                    }
                    adapter.submitList(response.body()?.books.orEmpty())
                }

                override fun onFailure(call: Call<SearchBookDTO>, t: Throwable) {
                    //실패처리
                    hideHistoryView()
                    Log.e(TAG, t.toString())
                }
            })
    }
    private fun initHistoryRecyclerView(){
        historyAdapter = HistoryAdapter(historyDeleteClickedListener = {
            deleteSearchKeyword(it)
        })
        binding.historyRecyclerView.layoutManager = LinearLayoutManager(this)
        binding.historyRecyclerView.adapter = historyAdapter
    }
    //삭제 후 뷰 갱신처리하기
    private fun showHistoryView(){
        Thread{
            val keywords = db.historyDao().getAll().reversed()
            //아래 코드가 없으면 저장은 되더라도 화면에 보이지 않아
            runOnUiThread{
                binding.historyRecyclerView.isVisible=true
                // 키워드에 널이 올 수 있으니까
                historyAdapter.submitList(keywords.orEmpty())
            }
        }.start()
    }
    //검색이 실패했을 때도 지워준다.
    private fun hideHistoryView(){
        binding.historyRecyclerView.isVisible = false
    }
    //검색된 키워드 등록처리
    private fun saveSearchKeyword(keyword: String){
        Thread{
            db.historyDao().insertHistory(History(null, keyword))
        }.start()
    }
    //검색 이력 삭제
    private fun deleteSearchKeyword(keyword: String){
        Thread{
            db.historyDao().delete(keyword)
            //삭제하고 나서 뷰 갱신처리하기
            showHistoryView()
        }.start()
    }

    companion object {
        private const val TAG = "MainActivity"
    }
}